Lambda functions, which have less code than normal functions but are more powerful when used properly, can be used in place of defining normal functions. Lambda functions are much simpler than normal functions.

Let us take a closer look at these Lamda functions.



In [ ]:
def double(x):
 return x*2

In [ ]:
[double, double, double]

[<function __main__.double(x)>,
 <function __main__.double(x)>,
 <function __main__.double(x)>]

The function double is defined in the first code. One feature of Python is its ability to specify variables as lists.

Just as with the variables, we can also declare the methods as a list.



In [ ]:
for func in [double, double, double]:
  print(func)

<function double at 0x7fe799c18280>
<function double at 0x7fe799c18280>
<function double at 0x7fe799c18280>


In [ ]:
number=1
for func in [double, double, double]:
  number=func(number)
  print(number)

2
4
8


First, we are providing an input of number=1. Then, three double functions are passed to the for loop. When the double method is passed first, the number doubles and is updated to 2. Next, in the second iteration, num 2 is supplied as an input to the double function. Next, it reaches four, at which point the number is updated and printed. The function receives the new number now, and it is updated to eight. After that, printing occurs.



In [ ]:
def double(x):
  return x * 2
def add_one(x):
  return x + 1

In [ ]:
number = 1
for func in [add_one, double, add_one]:
  number=func(number)
  print(number)

2
4
5


Comparably to the last example, we pass the add_one, double, and add_one methods as parameters rather than double, double, and double. This time, we are using num=1 and passing it as an argument to the func method, which updates the number to 2. The number two will then be printed as the initial output.

We will now take num=2 as input and provide it as a parameter to the double method, which will return num=4. After printing that, we will give the add_one method's input of num = 4 as input. After that, num is updated to 5. The value is then printed.


In [ ]:
number=1
for func in [lambda x:x + 1, lambda x: x * 2, lambda x: x * 4]:
  number=func(number)
  print(number)

2
4
16


We've redone the code's implementation. We have seen and implemented the identical logic here instead of decalving functions; the difference is that lambda functions have been employed this time. Here, we were able to use the same reasoning without even needing to specify the procedures. As you can see, it improved program performance and saved us a ton of lines.

This is the lambda function; no name with the same logic appears here. There is no name for this lambda function, therefore we can implement the logic here directly on the line.



In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4, 5]

reduce(lambda x, y: x + y, numbers)

15

We can see the actual potential of lambda functions with difficult and complex cases, rather than relying on such tiny examples. Here's the hard one to see now. The reduction module was first loaded from the functools library. We have the list of numbers 1, 2, 3, 4, and 5 that we have taken.

The logic now is to take the first two integers, x and y, add them, store the result as numbers = 3, then add them again with numbers = 2 (third iteration), and so on.

Numbers=3 and numbers=3; x=1, y=2, and x+y = 3. (from the enumeration)

numbers=6 and numbers= 4 (from list) and numbers=3 + 3 = 6.

numbers = 5 (from list), numbers = 10 and numbers = 6 + 4 = 10.

numbers=10 + 5 = 15, and the value will be printed if numbers=15.

Within the code "reduce(lambda x, y: x + y, numbers)" is what we're going to do. The lambda in this case represents the logic (in this case, the product), and the last argument is numbers, which is the data that the code will be sent.


In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4, 5]

reduce(lambda x, y: x * y, numbers)

120

We can see the actual potential of lambda functions with difficult and complex cases, rather than relying on such tiny examples. Here's the hard one to see now. The reduction module was first loaded from the functools library. We have the list of numbers 1, 2, 3, 4, and 5 that we have taken.

The logic now is to take the first two integers, x and y, add them, store the result as numbers = 3, then add them again with numbers = 2 (third iteration), and so on.

numbers=2 and numbers=3; x=1, y=2, and x*y = 2. (from the enumeration)

numbers=6 and numbers= 4 (from list) and numbers=2 * 3 = 6.

numbers = 24 and numbers = 5 (from list) and numbers = 6 * 4 = 24.

numbers=120 and the value will be printed if numbers=24 * 5 = 1120.

Within the code "reduce(lambda x, y: x + y, numbers)" is what we're going to do. The lambda in this case represents the logic (in this case, the product), and the last argument is numbers, which is the data that the code will be sent.



In [ ]:
import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.normal(0, 1, (10, 2)))
df.columns = ['column_a', 'column_b']
df.head()

column_a  column_b
0  1.500270 -0.637654
1 -0.765107  0.366100
2 -0.924019 -0.665384
3 -1.026763 -1.035801
4 -1.447379 -0.082665

It is more potent, particularly when used with the Pandas package.

Now let's examine how it is illustrated:

We just loaded the dataframe into the code above. We can now perform operations on it if we so choose. The pandas package makes it simple to accomplish.

Assume I wish to manipulate this data frame in some way. Assume for the moment that I wish to perform operations on the data that have positive values in column b.


In [ ]:
df.loc[lambda d: d['column_b'] > 0]


column_a  column_b
1 -0.765107  0.366100
5  0.706945  0.497192
6  0.075293  0.676230
8 -0.216518  0.378030

We are applying the filter to the dataframe while utilizing the loc technique. Only values are provided by this code, ensuring that all entries in the second column are positive.
